### ISSUE TITLE
Emoji support

### ISSUE THREAD

**Issue by exilonium** [prasanthrangan/hyprdots#285]
please add emoji keyboard support  
example :- https://github.com/linuxmobile/hyprland-dots/blob/Sakura/.config/hypr/keybinds.conf#L32
btw loved your project <3.

**Comment by katsuja** [comment]
Just install paru -S rofi-emoji and then put "bind = $mainMod, period, exec, wofi-emoji" in keybindings.conf

**Comment by piush-terminalwitchcraft** [comment]
It depends ! :( , either you have to remove rofi-lboon-wayland-git and install rofi (but it causes scaling issue in my hidpi screen)  because https://github.com/Mange/rofi-emoji doesnt works on wayland version.

**Comment by T-Crypt** [comment]
@exilonium 

This would most likely cause conflicts with the current install packages we are using for this repo. 

You can install this on your system though and test and then open a PR if you are able to mitigate the conflicts. I'll probably avoid messing with the conflicts between rofi package and the rofi-lboon-wayland-git!

**Comment by katsuja** [comment]
Do you mean something like that?
![231003_19h51m24s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/20012885/8884db21-6603-49ec-8231-1db41135119d)

**Comment by devCKVargas** [comment]
> please add emoji keyboard support example :- https://github.com/linuxmobile/hyprland-dots/blob/Sakura/.config/hypr/keybinds.conf#L32 btw loved your project <3.

Try `emote` from AUR

Add this to your keybindings.conf:
`bind = $mainMod, period, exec, emote # open emoji selector`

It's an alternative to rofi-emoji, it closes when you selected an emoji or hover away from it's window.
![emote](https://github.com/prasanthrangan/hyprdots/assets/50130192/e5cdb617-2c42-4501-b850-7bab668893b8)
![emote#2](https://github.com/prasanthrangan/hyprdots/assets/50130192/cae7213f-512b-47cc-b6dd-ebdce48a0f33)

**Comment by katsuja** [comment]
I use wofi-emoji and I have to say that I'm happy with it and it's not cumbersome
paru -S wofi-emoji
bind = $mainMod, period, exec, wofi-emoji 
add in keybindings.c
![231004_18h34m32s_screenshot](https://github.com/prasanthrangan/hyprdots/assets/20012885/2f3a0261-ab5e-4828-a5a7-e45e2f0ad0f1)
onf

**Comment by kRHYME7** [comment]
Hello, Remembered this thread. I use Emote, but for proof of concepts and more lightweight solution I added this feature on [hyprdots-ctl](https://github.com/kRHYME7/Hyprdots-ctl). 


Run: ``` Hyprdots show emoji ``` or add as a keybind 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/c022c65a-6ec1-40a5-87f2-cb8897acf8fa)



This uses the emoji database [here](https://github.com/Mange/emoji-data)  but also We can add custom emoji Data base.

**Comment by exilonium** [comment]
Thanks for this, but i've settled onto something more convenient wofi-emoji, since instead of only copying it also pastes back onto the textbox so i don't have to manually press ctrl+v , 

Could you bring out this functionality to it??

**Comment by exilonium** [comment]
i just installed wofi-emoji by `yay -S wofi-emoji` and set up the keybind `$mainMod , PERIOD, exec, wofi-emoji`

And it shows up like this.
![image](https://github.com/prasanthrangan/hyprdots/assets/146477733/d0cacf6c-3fc9-4da6-90d0-5f55ea68398f)

i type the emoji- name and press enter and it just copies it in clipboard and writes in the textbox. (just like they have it in windows)

**Comment by kRHYME7** [comment]
I see, best idea is sticking with wofi too. It's more compatible with Wayland. Maybe it's possible to port some rofi configuration to wofi.

**Comment by kRHYME7** [comment]
Hello @exilonium 
I recently added a default insert support that works similar to wofi-emoji 
command : ``` Hyprdots show emoji ``` 
also added experimental auto insert on ``` Hyprdots show pastebin ``` (clipboard) 


![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/f5dd4618-9660-48f8-b0cb-18cd9d885d4f)
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/73f1659f-0b18-4039-a270-84f269bb8e36)

update  ```hyprdots-ctl ```

https://github.com/kRHYME7/Hyprdots-ctl

**Comment by exilonium** [comment]
Thank you so much it's now working pretty good i've switched to it 😀

**Comment by exilonium** [comment]
This works nicely but i think this is too big on screen , can we have it of the size of pastebin and other stuff and also make it follow the mouse like pastebin do??
and also i got confused since pasteboard was switched to pastebin  

btw i'm not getting the tick emoji (i've jetbrains and cascadia both nerd fonts installed)

**Comment by kRHYME7** [comment]
I think rofi has keybind for marking. It's shift + enter

**Comment by exilonium** [comment]
thanks for it now it works ...
A feature i've been missing is that emoji keyboard doesn't update our recently used emoji and give it high preference.
let's say i use heart emoji the most, searching for heart brings this emoji at top 🥰 but i use ️️❤️, even if i use this ❤️ emoji the preferences will still show 🥰 emoji at top....

**Comment by kRHYME7** [comment]
wofi-emoji have this? is so maybe I can check their source. 
But for now I can only offer a "recent" feature which just caching the last selected emoji at the top. 

In theory if you search the heart keyword the recently selected  emoji with the heart keyword will appear on top. We can't really use rofi for much more complex execution as this is not designed for this.

**Comment by exilonium** [comment]
yup that would be appreciated too..maybe we can store the used emoji in a database or csv and store how many times the emoji have been used and use this data to manipulate the algorithm ??

**Comment by kRHYME7** [comment]
There's not much of a logic when dealing with  rofi though. It just lists whatever you feed to it. 
I have some approach in my head right now.

1. Copy the whole emoji.db then re order by most recently used. (slow, won't receive updated emoji.db)

2.0 Make rofi read 2 files the main emoji.db and the recent-emoji.db. Rofi first read the recent-emoji.db displaying at the top most.  Now we can update the emoji.db upstream and we have the recent.

Now dealing with Duplicates.
2.1 Just ignore the duplicates it won't hurt and makes the whole execution faster. (The recent-emoji.db )

2.2 Limit the recent-emoji. to about maybe 5-10  recent emoji. with this we won't have a big number of duplicates. (Emoji on phone keyboards somewhat acts like this)

2.3 Add a deduplication function that removes the same emoji lines from displaying (Slow but cool 😎,) 
    2.3.a won't be slow if we only de duplicate the Recent-emoji

     2.4.b Slow if we deduplicate all the 1848 lines of emoji.



I'll try to implement 2.3.a . What else might you think of?

**Comment by kRHYME7** [comment]
There's not much of a logic when dealing with  rofi though. It just lists whatever you feed to it. 
I have some approach in my head right now.

1. Copy the whole emoji.db then re order by most recently used. (slow, won't receive updated emoji.db)

2.0 Make rofi read 2 files the main emoji.db and the recent-emoji.db. Rofi first read the recent-emoji.db displaying at the top most.  Now we can update the emoji.db upstream and we have the recent.

Now dealing with Duplicates.
2.1 Just ignore the duplicates it won't hurt and makes the whole execution faster. (The recent-emoji.db )

2.2 Limit the recent-emoji. to about maybe 5-10  recent emoji. with this we won't have a big number of duplicates. (Emoji on phone keyboards somewhat acts like this)

2.3 Add a deduplication function that removes the same emoji lines from displaying (Slow but cool 😎,) 
    2.3.a won't be slow if we only de duplicate the Recent-emoji

     2.4.b Slow if we deduplicate all the 1848 lines of emoji.



I'll try to implement 2.3.a . What else might you think of?

**Comment by kRHYME7** [comment]
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/97ce86ff-d5b5-4dd5-9f74-4fd881fcf178)

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/cea8fe67-3421-4397-bc27-af381c0ebe5d)


 Solved the duplicates.
 I implemented  ``` 2.4.b Slow if we deduplicate all the 1848 ++ lines of emoji. ```  And it's fast.
  
 Dunno how fast Yay/paru updates git but I recently pushed a recent commit 
 
🛞🇦🇩🇦🇪🇦🇫🇦🇬🇦🇮🇦🇱🇦🇲🇦🇴🇦🇶🇦🇷🇦🇸🇦🇹🇦🇺🇦🇼🇦🇽🇦🇿🇧🇦🇧🇧🇧🇩🇧🇪🇧🇫🇧🇬🇧🇭🇧🇮🇧🇯🇧🇱🇧🇲🇧🇳🇧🇴🇧🇶🇧🇷🇧🇸🇧🇹🇧🇼🇧🇾🇧🇿🇨🇦🇨🇨🇨🇩🇨🇫🇨🇬🇨🇭🇨🇮🇨🇰🇨🇱🇨🇲🇨🇳🇨🇴🇨🇷🇨🇺🇨🇻🇨🇼🇨🇽🇨🇾🇨🇿🇩🇪🇩🇯🇩🇰🇩🇲🇩🇴🇩🇿🇪🇨🇪🇪🇪🇬🇪🇭🇪🇷🇪🇸🇪🇹🇫🇮🇫🇯🇫🇰🇫🇲🇫🇴🇫🇷🇬🇦🇬🇧🇬🇩🇬🇪🇬🇫🇬🇬🇬🇭🇬🇮🇬🇱🇬🇲🇬🇳🇬🇵🇬🇶🇬🇷🇬🇸🇬🇹🇬🇺🇬🇼🇬🇾🇭🇰🇭🇳🇭🇷🇭🇹🇭🇺🇮🇨🇮🇩🇮🇪🇮🇱🇮🇲🇮🇳🇮🇴🇮🇶🇮🇷🇮🇸🇮🇹🇯🇪🇯🇲🇯🇴🇯🇵🇰🇪🇰🇬🇰🇭🇰🇮🇰🇲🇰🇳🇰🇵🇰🇷🇰🇼🇰🇾🇰🇿🇱🇦🇱🇧🇱🇨🇱🇮🇱🇰🇱🇷🇱🇸🇱🇹🇱🇺🇱🇻🇱🇾🇲🇦🇲🇨🇲🇩🇲🇪🇲🇬🇲🇭🇲🇰🇲🇱🇲🇲🇲🇳🇲🇴🇲🇵🇲🇶🇲🇷🇲🇸🇲🇹🇲🇺🇲🇻🇲🇼🇲🇽🇲🇾🇲🇿🇳🇦🇳🇨🇳🇪🇳🇫🇳🇬🇳🇮🇳🇱🇳🇴🇳🇵🇳🇷🇳🇺🇳🇿🇴🇲🇵🇦🇵🇪🇵🇫🇵🇬🇵🇭🇵🇰🇵🇱🇵🇲🇵🇳🇵🇷🇵🇸🇵🇹🇵🇼🇵🇾🇶🇦🇷🇪🇷🇴🇷🇸🇷🇺🇷🇼🇸🇦🇸🇧🇸🇨🇸🇩🇸🇪🇸🇬🇸🇭🇸🇮🇸🇰🇸🇱🇸🇲🇸🇳🇸🇴🇸🇷🇸🇸🇸🇹🇸🇻🇸🇽🇸🇾🇸🇿🇹🇨🇹🇩🇹🇫🇹🇬🇹🇭🇹🇯🇹🇰🇹🇱🇹🇲🇹🇳🇹🇴🇹🇷🇹🇹🇹🇻🇹🇼🇹🇿🇺🇦🇺🇬🇺🇸🇺🇾🇺🇿🇻🇦🇻🇨🇻🇪🇻🇬🇻🇮🇻🇳🇻🇺🇼🇫🇼🇸🇽🇰🇾🇪🇾🇹🇿🇦🇿🇲🇿🇼

**Comment by exilonium** [comment]
Thank you so much IDK much what's going on in backend but it seems to work pretty well.

BUT now it doesn't pastes emoji's automatically , it just copies in clipboard and doesn't paste

**Comment by kRHYME7** [comment]
> BUT now it doesn't pastes emoji's automatically , it just copies in clipboard and doesn't paste

Thanks! should fix some errors now. 
Can I ask a favor about testing it on Terminal, the xwayland apps, and Wayland app?

 You might get inconsistent output  in Wayland and Xwayland windows

**Comment by exilonium** [comment]
i mostly use emoji on internet so i tried it in brave-browser an it didn't worked

**Comment by kRHYME7** [comment]
have you updated hyprdots-ctl?

**Comment by kRHYME7** [comment]
I see, what is the output for ``` pacman -Q wtype ```

**Comment by exilonium** [comment]
> pacman -Q wtype

wtype 0.4-1

**Comment by kRHYME7** [comment]
Strange... can you try to ``` paru -Sy hyprdots-ctl --rebuild ``` again 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/6658d709-f41d-443a-b5c5-799e09d5d874)

It works perfectly on my machine so I will try to find another system to test.

**Comment by exilonium** [comment]
Yes now it's working fine 😀 (<--this emoji is pasted using your tool :D )

**Comment by kRHYME7** [comment]
Feel free to post an issue here https://github.com/kRHYME7/Hyprdots-ctl/issues. So we can keep track of the progress.
